lightning-c-bindings/src/c_types/mod.rs

```diff
@@ -0,0 +1,415 @@
+			datalen: s.len(),
+		}
+	}
+}
+
+#[repr(C)]
+#[derive(Clone)]
+/// A transaction output including a scriptPubKey and value.
+/// This type *does* own its own memory, so must be free'd appropriately.
+pub struct TxOut {
+	pub script_pubkey: derived::CVec_u8Z,
+	pub value: u64,
+}
+
+impl TxOut {
+	pub(crate) fn into_rust(self) -> ::bitcoin::blockdata::transaction::TxOut {
+		::bitcoin::blockdata::transaction::TxOut {
+			script_pubkey: self.script_pubkey.into_rust().into(),
+			value: self.value,
+		}
+	}
+	pub(crate) fn from_rust(txout: ::bitcoin::blockdata::transaction::TxOut) -> Self {
+		Self {
+			script_pubkey: CVecTempl::from(txout.script_pubkey.into_bytes()),
+			value: txout.value
+		}
+	}
+}
+#[no_mangle]
+pub extern "C" fn TxOut_free(_res: TxOut) { }
+
+#[repr(C)]
+pub struct u8slice {
+	pub data: *const u8,
+	pub datalen: usize
+}
+impl u8slice {
+	pub(crate) fn from_slice(s: &[u8]) -> Self {
+		Self {
+			data: s.as_ptr(),
+			datalen: s.len(),
+		}
+	}
+	pub(crate) fn to_slice(&self) -> &[u8] {
+		if self.datalen == 0 { return &[]; }
+		unsafe { std::slice::from_raw_parts(self.data, self.datalen) }
+	}
+}
+
+#[repr(C)]
+pub struct usizeslice {
+	pub data: *const usize,
+	pub datalen: usize
+}
+impl usizeslice {
+	pub(crate) fn from_slice(s: &[usize]) -> Self {
+		Self {
+			data: s.as_ptr(),
+			datalen: s.len(),
+		}
+	}
+	pub(crate) fn to_slice(&self) -> &[usize] {
+		if self.datalen == 0 { return &[]; }
+		unsafe { std::slice::from_raw_parts(self.data, self.datalen) }
+	}
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+/// Arbitrary 32 bytes, which could represent one of a few different things. You probably want to
+/// look up the corresponding function in rust-lightning's docs.
+pub struct ThirtyTwoBytes {
+	pub data: [u8; 32],
+}
+impl ThirtyTwoBytes {
+	pub(crate) fn empty() -> Self {
+		Self { data: [0; 32] }
+	}
+}
+
+#[repr(C)]
+pub struct ThreeBytes { pub data: [u8; 3], }
+#[derive(Clone)]
+#[repr(C)]
+pub struct FourBytes { pub data: [u8; 4], }
+#[derive(Clone)]
+#[repr(C)]
+pub struct TenBytes { pub data: [u8; 10], }
+#[derive(Clone)]
+#[repr(C)]
+pub struct SixteenBytes { pub data: [u8; 16], }
+
+pub(crate) struct VecWriter(pub Vec<u8>);
+impl lightning::util::ser::Writer for VecWriter {
+	fn write_all(&mut self, buf: &[u8]) -> Result<(), ::std::io::Error> {
+		self.0.extend_from_slice(buf);
+		Ok(())
+	}
+	fn size_hint(&mut self, size: usize) {
+		self.0.reserve_exact(size);
+	}
+}
+pub(crate) fn serialize_obj<I: lightning::util::ser::Writeable>(i: &I) -> derived::CVec_u8Z {
+	let mut out = VecWriter(Vec::new());
+	i.write(&mut out).unwrap();
+	CVecTempl::from(out.0)
+}
+pub(crate) fn deserialize_obj<I: lightning::util::ser::Readable>(s: u8slice) -> Result<I, lightning::ln::msgs::DecodeError> {
+	I::read(&mut s.to_slice())
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+/// A Rust str object, ie a reference to a UTF8-valid string.
+/// This is *not* null-terminated so cannot be used directly as a C string!
```

Comment:
Why not just use C strings when interacting with C?

Reply:
Hmm, in general I think the theory of the bindings here is that they should target being "just the Rust code" as much as practical - at the end of the day its probably not being used from C and will take another hop to get to Java/JavaScript/Swift/etc so might as well let the next hop do the real conversion, instead of assuming what they want.